### DISCUSSION TITLE
SECURITY: Is the csrf cookie enough to be safe against the CSRF attack? Shouldn't we also send a header in the call? Isn't that the CSRF point?

### DISCUSSION THREAD

**Discussion by frederikhors** [ory/kratos#1525] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

I'm using Kratos and Oathkeeper: what a joy!

I have a SPA (single page app).

The `csrf` and `session` cookie are there.

I call oathkeeper with those cookies only and it authenticates me against `Kratos` like well explained [here](https://www.ory.sh/kratos/docs/guides/zero-trust-iap-proxy-identity-access-proxy).

**Describe the solution you'd like**

Should we send within each (`POST` or `GET`) call an header like `X-CSRF-Token` or something similar?

I think the `csrf` cookie is not enough for CSRF security attack (from what I understand by studying the problem).

Am I wrong?

**Comment by frederikhors** [comment]
More on this here: https://stackoverflow.com/a/34783845

> An alternative approach (called the "Cookie-to-header token" pattern) is to set a Cookie once per session and the have JavaScript read that cookie and set a custom HTTP header (often called X-CSRF-TOKEN or X-XSRF-TOKEN or just XSRF-TOKEN) with that value. Any requests will send both the header (set by Javascript) and the cookie (set by the browser as a standard HTTP header) and then the server can check that value in the X-CSRF-TOKEN header matches the value in the cookie header. The idea being that only JavaScript run on the same domain would have access to the cookie, so JavaScript from another domain couldn't set this header to the right value (assuming the page is not vulnerable to XSS that would give access to this cookie). Even fake links (e.g. in a phishing email) would not work either, as even though they would appear to come from the right domain, only the cookie will be set but not X-CSRF-TOKEN header.

https://en.wikipedia.org/wiki/Cross-site_request_forgery#Cookie-to-header_token

**Comment by frederikhors** [comment, marked as answer]
@aeneasr Why are you closing it? Isn't that a serious security issue?

**Reply by aeneasr** [reply]
No, it’s a question and not a technical issue so belongs it here.

**Reply by frederikhors** [reply]
Ok. I opened the issue because I think it is a security problem if the CSRF is not managed correctly.

**Reply by aeneasr** [reply]
If you think you found a security issue please follow the repository’s security policy: https://github.com/ory/kratos/security

However I’m very confident you have not found such an issue. But the report is missing tons of detail (for example which API you think is affected, a reproducible attack, …) so I might be wrong

**Reply by frederikhors** [reply]
I think all APIs are affected by this problem, not just a few.

The CSRF cookie alone IS NOT ENOUGH to solve the problem, we need more, as explained here: https://stackoverflow.com/a/34783845.

If I send an email to security@ory.sh it's the same. You can use this as an open issue to learn more about CSRF since you are about to release the new version for AJAX.

Please, I'm not a bother, I love the product you are making and I want to help. I'm not here to waste mine and yours time.

Reconsider this. Thanks.

**Reply by aeneasr** [reply]
All self-service endpoints have appropriate CSRF protection. The mechanisms are well documented and have tests. And of course they do not rely on just the cookie being there. Admin endpoints have no CSRF protection but these are not facing browsers and the public internet (if they do in your system you have other concerns than CSRF). For AJAX and others we also already support the X-CSRF-Token Header wherever applicable and sensible. Also endpoints which do not modify resources (such as fetching your session) do not require CSRF protection as these are protected using different mechanisms such as cookie domain, httpOnly and others.

To me it sounds right now as if there are some misunderstandings how the APIs work and also potentially some oversights in the payloads which, again, always require the CSRF token to be sent ad part of the payload and the cookie (for validation).

So please, include details and narrow down what endpoints you’re talking about specifically, include examples, and some reproducible proof that they are actually vulnerable to CSRF. Otherwise this discussion will go nowhere. One thing I can say with absolute certainty is that all appropriate endpoints have security measures applied where necessary. The only thing that could happen is that one endpoint has some type of edge caee where the protection fails.

If you find such a vulnerability please disclose it according to our security policy and not publicly on GitHub.

Thank you for your understanding and have a great Sunday :)

**Reply by frederikhors** [reply]
Ok. I'll follow security policy.

Just one thing: let's say I'm using grpc (they are all POST requests).

Using Oathkeeper and Kratos with `cookie_session` method I'm able to talk with my backend grpc server after the login (with `flow/browser`) with session cookie only.

I think this is wrong because Kratos should detect I'm using the browser and should ask for an header (as reported in that link I posted before).

Do you think I'm wrong about this too?

**Reply by aeneasr** [reply]
I don't know of any browser that supports gRPC. CSRF is only an issue in browser environments, so yes, I don't think it's an issue there. But I don't know your application, so who knows ;) If you feel that your app needs CSRF protection, add it to your app.

Ory Kratos only protects its own endpoints, not yours, against CSRF.

**Reply by frederikhors** [reply]
There is [grpc-web](https://github.com/improbable-eng/grpc-web).

But also GraphQL, for example, or any other backend that supports POST calls.

And anyway I think it is Kratos' job to manage this, not my application. Basically **this is authentication**.

And if I use Oathkeeper it is precisely because I want to keep the scopes separate.

A CSRF working for Kratos' public API but not for the application is like having no authentication at all.

IMHO.

Please don't get mad.

**Reply by frederikhors** [reply]
Can we add a way to check for CSRF header too?

**Reply by aeneasr** [reply]
No it really isn’t Ory Kratos‘ job nor is it possible to implement this. Ory Kratos doesn’t even speak GRPC. You need to solve this in your application. CSRF tokens are something to protect you from cross site request forgery. Authentication is letting a system know who you are. These are completely different things and it’s absolutely out of scope to be implemented in Ory Kratos.

**Reply by aeneasr** [reply]
I’m not mad by the way ;) It just wasn‘t clear what you wanted. Now that I understand that you want to protect your application from CSRF, and have not found an issue in Ory Kratos, it is clear what you wanted to say i. the original issue.

It’s great that you’re a fan! Sometimes these discussions (on a Sunday ;)) can stress out maintainers - sorry if the tone came on as harsh. I guess that’s the germaness in me ;)

Anyways, there really isn’t a way to do what you want for Ory Kratos and it’s not in the scope of the project.

However, adding CSRF protections is really easy as there are many libraries available that you can just include!

If you do find a CSRF issue in Ory Kratos itself please report it to the appropriate security channel. Thank you!

**Reply by frederikhors** [reply]
I understand what you meant. Thanks for the calm and fruitful response.

You are doing a wonderful job with Ory!

You are legends! Thanks so much!

**Reply by aeneasr** [reply]
Thank you @frederikhors :)
